osconfig_tests/test_suites/package_management/package_management_test_data.go

```diff
@@ -0,0 +1,169 @@
+		}
+		setup := packageManagementTestSetup{
+			image:      debianImage,
+			name:       instancename,
+			osconfig:   oc,
+			assignment: assign,
+			fname:      testName,
+			vf:         vf,
+			startup: &api.MetadataItems{
+				Key:   "startup-script",
+				Value: &utils.StartupScriptDeb,
+			},
+		}
+		pkgTestSetup = append(pkgTestSetup, &setup)
+	}
+}
+
+func addPackageRemovalTest() {
+	testName := "packageremovaltest"
+	desc := "test package removal"
+	for _, pkgManager := range pkgManagers {
+		var oc *osconfigpb.OsConfig
+		var assign *osconfigpb.Assignment
+		var instancename string
+
+		switch pkgManager {
+		case "apt":
+			instancename = fmt.Sprintf("%s-%s", filepath.Base(debianImage), testName)
+			pkg := osconfigserver.BuildPackage("cowsay")
+			pkgs := []*osconfigpb.Package{pkg}
+			oc = osconfigserver.BuildOsConfig(testName, desc, osconfigserver.BuildAptPackageConfig(nil, pkgs, nil), nil, nil, nil, nil)
+			assign = osconfigserver.BuildAssignment(testName, desc, osconfigserver.BuildInstanceFilterExpression(instancename), []string{fmt.Sprintf("projects/%s/osConfigs/%s", testProjectID, oc.Name)})
+		}
+		setup := packageManagementTestSetup{
+			image:      debianImage,
+			name:       instancename,
+			osconfig:   oc,
+			assignment: assign,
+			fname:      testName,
+			vf:         vf,
+			startup: &api.MetadataItems{
+				Key:   "startup-script",
+				Value: &utils.StartupScriptDeb,
+			},
+		}
+		pkgTestSetup = append(pkgTestSetup, &setup)
+	}
+}
+
+func addPackageInstallRemovalTest() {
+	testName := "packageinstallremovaltest"
+	desc := "test package removal takes precedence over package installation"
+	for _, pkgManager := range pkgManagers {
+		var oc *osconfigpb.OsConfig
+		var assign *osconfigpb.Assignment
+		var instancename string
+
+		switch pkgManager {
+		case "apt":
+			instancename = fmt.Sprintf("%s-%s", filepath.Base(debianImage), testName)
+			pkg := osconfigserver.BuildPackage("cowsay")
+			installPkg := []*osconfigpb.Package{pkg}
+			pkg = osconfigserver.BuildPackage("cowsay")
+			removePkg := []*osconfigpb.Package{pkg}
+			oc = osconfigserver.BuildOsConfig(testName, desc, osconfigserver.BuildAptPackageConfig(installPkg, removePkg, nil), nil, nil, nil, nil)
+			assign = osconfigserver.BuildAssignment(testName, desc, osconfigserver.BuildInstanceFilterExpression(instancename), []string{fmt.Sprintf("projects/%s/osConfigs/%s", testProjectID, oc.Name)})
+		}
+		setup := packageManagementTestSetup{
+			image:      debianImage,
+			name:       instancename,
+			osconfig:   oc,
+			assignment: assign,
+			fname:      testName,
+			vf:         vf,
+			startup: &api.MetadataItems{
+				Key:   "startup-script",
+				Value: &utils.StartupScriptDeb,
+			},
+		}
+		pkgTestSetup = append(pkgTestSetup, &setup)
+	}
+}
+
+func generateAllTestSetup() []*packageManagementTestSetup {
```

Comment:
Are these test "setups" or the actual tests?

Reply:
my idea of "setup" was creating any resource(and thus the environment) to run a particular test. I wanted these setups to be de-coupled from the actual tests. but the setup is really tied the actual tests in the sense that for a particular test, there is a particular environment required.
that's the reason i had to introduce the "fname" in the setup struct and do all the filtering logic in the test runner.
honestly, i wanted to structure this in a better way so that the test runner did not have to do the filtering based on the function name( test runner need not have to have any context about the testsetup). but since i had already spent too much time figuring this out, in the interest of time, i just went ahead with my current logic.